### ISSUE TITLE
LayerNormalization 的计算

### ISSUE THREAD

**Issue by i4never** [bojone/bert4keras#318]
https://github.com/bojone/bert4keras/blob/eb7932a1ea023c641ebc180c49136fa320a39bbb/bert4keras/layers.py#L471

```python
    @recompute_grad
    def call(self, inputs):
        """如果是条件Layer Norm，则默认以list为输入，第二个是condition
        """
        if self.conditional:
            inputs, cond = inputs
            if self.hidden_units is not None:
                cond = self.hidden_dense(cond)
            for _ in range(K.ndim(inputs) - K.ndim(cond)):
                cond = K.expand_dims(cond, 1)
            if self.center:
                beta = self.beta_dense(cond) + self.beta
            if self.scale:
                gamma = self.gamma_dense(cond) + self.gamma
        else:
            if self.center:
                beta = self.beta
            if self.scale:
                gamma = self.gamma

        outputs = inputs
        if self.center:
            mean = K.mean(outputs, axis=-1, keepdims=True)
            outputs = outputs - mean
        if self.scale:
            variance = K.mean(K.square(outputs), axis=-1, keepdims=True)
            std = K.sqrt(variance + self.epsilon)
            outputs = outputs / std
            outputs = outputs * gamma
        if self.center:
            outputs = outputs + beta

        return outputs
```

- 1. 在`center=False`，`scale=True`时的计算可能有问题，outputs的方差计算没有减去均值。
- 2. 对比了tf的LayerNormalization（keras.layers.LayerNormalization）计算，无论center/scale与否，对于输入始终做了减去均值除以方差的操作（除以方差，减去均值除以方差），只是在scale时，额外乘了weights；在center时额外加了bias，与bert4keras中的实现有区别。

如果照搬过来并解决问题1，call的写法是

```python
    @recompute_grad
    def call(self, inputs):
        """如果是条件Layer Norm，则默认以list为输入，第二个是condition
        """
        if self.conditional:
            inputs, cond = inputs
            if self.hidden_units is not None:
                cond = self.hidden_dense(cond)
            for _ in range(K.ndim(inputs) - K.ndim(cond)):
                cond = K.expand_dims(cond, 1)
            if self.center:
                beta = self.beta_dense(cond) + self.beta
            if self.scale:
                gamma = self.gamma_dense(cond) + self.gamma
        else:
            if self.center:
                beta = self.beta
            if self.scale:
                gamma = self.gamma

        outputs = inputs

        mean = K.mean(outputs, axis=-1, keepdims=True)
        variance = K.mean(K.square(outputs - mean), axis=-1, keepdims=True)

        inv = 1. / K.sqrt(variance + self.epsilon)

        if self.scale:
            inv *= gamma

        outputs = outputs * inv - mean * inv

        if self.center:
            outputs += beta

        return outputs
```

**Comment by bojone** [comment]
这...我也没说center和scale的含义跟官方的一样啊，我就是要实现我现在实现的功能啊......

**Comment by i4never** [comment]
但是T5模型，只scale没有center，和预训练的流程就不同了是吗？差了一个bias

**Comment by bojone** [comment]
> 但是T5模型，只scale没有center，和预训练的流程就不同了是吗？差了一个bias

何以见得？T5预训练的时候就没有减均值的好吧。

**Comment by i4never** [comment]
https://github.com/tensorflow/mesh/blob/441ff477dcff8fc577eb59019007e55a278c87d6/mesh_tensorflow/transformer/transformer.py#L492
抱歉、做的是RMS normalization、是我疏忽了、

**Comment by battleman123** [comment]
> > 但是T5模型，只scale没有center，和预训练的流程就不同了是吗？差了一个bias
> 
> 何以见得？T5预训练的时候就没有减均值的好吧。

请问用bert4keras训练好了一个版本的权值，现在想基于huggingface实现的t5加载训练好的权值继续训练，背景是这个。
同样的输入，两个模型的输出是不一样的，找到一个原因是LayerNorm方式不一样，这样不就不能基于bert4keras训练出来的权值在huggingface上的T5继续训练了？有没有人这样尝试过呢？
